x-pack/legacy/plugins/apm/public/components/app/TransactionDetails/WaterfallWithSummmary/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -277,8 +279,10 @@ export function getWaterfall(
   const services = getServices(orderedItems);
   const getTransactionById = createGetTransactionById(itemsById);
   const serviceColors = getServiceColors(services);
+  const entryTransaction = getTransactionById(entryTransactionId);
 
   return {
+    entryTransaction,
```

Comment:
Probably out of scope for this PR, but this file is hard to understand for me. Not just the file itself but also the data structure that it returns. It's hard to tell just by looking at this how it will be used. I also wonder if it belongs on the server?

Reply:
I completely agree - it's pretty confusing atm. Moving most of this to the backend, so just the relevant part of the waterfall is returned (and not the entire trace) will probably be a good start in cleaning this up.

We might also at the same time want to look into lazy loading the waterfall instead of just the first 1000. Felix had some comments on how that could be achieved: https://github.com/elastic/kibana/issues/20660#issuecomment-417380323

Reply:
But yeah - it's out-of-scope for this PR. I actually started on it but it became a bigger task than I had anticipated.